### DISCUSSION TITLE
Add buildin artifact support

### DISCUSSION THREAD

**Discussion by kyngs** [woodpecker-ci/woodpecker#2282] [Request a Feature]
### Clear and concise description of the problem

I have a pipeline which builds my project. I want to attach the JAR files it builds to the pipeline.

### Suggested solution

It would work similarly to Jenkins build artifacts.
Perhaps in this section:
![image](https://user-images.githubusercontent.com/38181667/216778398-b81f72de-09e4-4c16-bea9-e380d7242c40.png)
a tab named something like "Build Artifacts" could be added.
The tab should contain the files, which the viewer could download.

### Alternative

Adding a step that would upload the artifacts to an external server.

### Additional context

I couldn't find any issues on the same subject, which seems surprising considering that this is a pretty popular use case.
Maybe someone already suggested it, and it was rejected? I don't know.

### Validations

- [x] Read the [Contributing Guidelines](https://github.com/woodpecker-ci/woodpecker/blob/master/CONTRIBUTING.md).
- [X] Read the [docs](https://woodpecker-ci.org/docs/intro).
- [X] Check that there isn't already an [issue](https://github.com/woodpecker-ci/woodpecker/issues) that request the same feature to avoid creating a duplicate.

**Comment by 6543** [comment]
Must be optional and should only be available if an s3 storage option was provided to woodpecker

**Comment by kyngs** [comment]
> Must be optional and should only be available if an s3 storage option was provided to woodpecker

My original idea was, that it would be stored on the woodpecker itself or some other self-hosted solution.

**Comment by neel** [comment]
@kyngs gitlab-runner has artifacts. I was looking for something like that in Woodpecker. It doesn't require S3. Are you also requesting for similar feature ?

**Comment by kyngs** [comment]
> @kyngs gitlab-runner has artifacts. I was looking for something like that in Woodpecker. It doesn't require S3. Are you also requesting for similar feature ?

Yes, Jenkins also has this feature, and so do Github actions. It is really useful for sharing binaries for each commit.

**Comment by neel** [comment]
> > @kyngs gitlab-runner has artifacts. I was looking for something like that in Woodpecker. It doesn't require S3. Are you also requesting for similar feature ?
> 
> Yes, Jenkins also has this feature, and so do Github actions. It is really useful for sharing binaries for each commit.

Not only that, it is very useful for writing papers in Latex. For every push the pdf is generated automatically.

I think Kraken CI also has this feature.

**Comment by aozomaro** [comment]
I have built a solution for this in our project.

- I’ve created a plugin called save_artifacts. This follows very closely with the save_artifacts paradigm used by circleci.
- The plugin only accepts the mount location and uses internally set environment variables (CI_Artifacts_host, etc) 
- Currently the plugin only persists to an s3 like store, however for this proposal I would re use drone-cache which would allow more flexibility in object store type (azure, sftp, etc)
- I’ve added a tab to the pipeline.vue page to display artifacts
- I’ve added a v2 api to retrieve artifacts (it uses woodpecker api to authenticate the user)

there’s additional processes which are work in progress to control the retention time of artifacts

I would like to start by adding in the api to retrieve, configuration options (similar to the private docker repositories config) the plugin to store artifacts and the tab. 

My code is ready so if the proposal above seems good I can start pushing up some PRs!

**Reply by qwerty287** [reply]
@aozomaro Feel free to suggest PRs - we'll review it!
The proposal itself looks like it's mainly about linking an artifact in the UI, but not storing it. Is this correct? (So your approach would only solve https://github.com/woodpecker-ci/woodpecker/discussions/2298?) This issue/discussion is really about "native" artifacts which includes storing them on the server and not only linking them.
If the code is already there, just open a PR and we can continue discussing and reviewing your changes there.

**Comment by zc-devs** [comment]
> that it would be stored on the woodpecker itself or some other self-hosted solution.

It should be stored in special artifact repository: [Nexus](https://www.sonatype.com/products/sonatype-nexus-oss), [Artifactory](https://jfrog.com/artifactory/), [Gitea](https://docs.gitea.com/usage/packages/overview), etc in your JAR case.

> I want to attach the JAR files it builds to the pipeline.

If replace "attach" to "list", then make separate step and list whatever you want, like other tools do. Or do it in the end of build step / upload step / whatever step you want.

1. What is `the artifact`? Is security check report artifact? Tests execution report? Docker image?
2. Should we copy-paste [this functionality](https://docs.gitea.com/usage/packages/overview)?

**Reply by qwerty287** [reply]
I think we can easily solve this by ui plugins/linking artifacts in the ui.
We can just say that we won't implement native file storage in woodpecker, but you can use plugins and then link the files.

**Reply by kyngs** [reply]
I imagine this would work similarly to GitHub actions where you can use a pipeline part to upload the build artifacts to GitHub (in this case I imagine it would be uploaded to Woodpecker) and then when you view the relevant build overview you can download the artifacts.

I definitely do not want to upload it to a 3rd party solution

**Reply by qwerty287** [reply]
> I definitely do not want to upload it to a 3rd party solution

Why? Yes, I see it can add more complexity to your setup of you self-host the file storage, but we can't support all use cases as @zc-devs said already. Of course we can say to only allow file artifacts, but what about other "formats" (yes, everything is a file ;) ) like package registries?

**Reply by kyngs** [reply]
> Yes, I see it can add more complexity to

That's the issue. I think there's a good reason why most other CI systems support this.

> Of course we can say to only allow file artifacts, but what about other "formats"

I'm not sure I understand what you mean here? I intended this for simple build artifacts publication such as JAR files, executable files etc.

Would a simple file storage be hard to implement to woodpecker?


I understand that woodpecker doesn't want to get bloated, but I believe this is a basic feature.
